p2p/tpu.md

```diff
@@ -0,0 +1,234 @@
+
+**Example QUIC frames**
+
+_First UDP packet_
+
+```
+STREAM Frame {
+  Prefix (5) = 0b00001
+  Offset Present (1) = 0   # Offset is absent for first fragment
+  Length Present (1) = 0   # Length is absent, as fragment frames always fill packet
+  Fin (1) = 0
+
+  Stream ID (i)
+  Stream Data (..)
+}
+```
+
+_Second UDP packet_
+
+```
+STREAM Frame {
+  Prefix (5) = 0b00001
+  Offset Present (1) = 1
+  Length Present (1) = 0
+  Fin (1) = 0
+
+  Stream ID (i)
+  [Offset (i)]
+  Stream Data (..)
+}
+```
+
+_Last UDP packet_
+
+```
+STREAM Frame {
+  Prefix (5) = 0b00001
+  Offset Present (1) = 1
+  Length Present (1) = 0
+  Fin (1) = 1             # Terminate stream
+
+  Stream ID (i)
+  [Offset (i)]
+  Stream Data (..)
+}
+```
+
+#### Acknowledgement
+	
+`STREAM` frames sent by clients are ack-eliciting.
+Refer to [RFC 9000 Section 13.2.1] for obligations on sending `ACK` frames.
+
+It is permitted to coalesce multiple `ACK` frames to reduce packet count.
+
+#### Stream Limits
+
+The server should send [`MAX_STREAM_DATA`] frames for recently created fragmented tx streams.
+The client should expect arbitrarily receive quotas for any stream.
+
+Note that stream IDs are incremental (ignoring the two least significant bits).
+Upon opening a stream with an out-of-order ID, all preceding streams are also created.
+
+In this case, the server may proactively send multiple [`MAX_STREAM_DATA`] frames for implicitly created streams.
+
+  [RFC 9000 Section 13.2.1]: https://www.rfc-editor.org/rfc/rfc9000.html#name-sending-ack-frames
+  [`MAX_STREAM_DATA`]: https://www.rfc-editor.org/rfc/rfc9000.html#name-max_stream_data-frames
+
+#### Aborting
+
+The server must terminate the QUIC connection in case of the following events:
+
+  - Submission of invalid transactions, specifically:
+    - Txs that fail deserialization
+    - Txs that fail parameter sanitization
+    - Txs that fail signature verification
+  - Transmission of QUIC datagrams, defined by [RFC 9221]
+  - Attempted creation of QUIC bidirectional streams (already forbidden by quota)
+  - All other cases where the QUIC RFC dictates connection termination.
+
+  [RFC 9221]: https://www.rfc-editor.org/rfc/rfc9221.html
+
+## TPU/UDP protocol
+
+TPU/UDP is a simple datagram-oriented network protocol.
+
+Each datagram carries a single serialized transaction without headers or padding.
+
+It is being deprecated in favor of TPU/QUIC, which adds authentication, confidentiality, and congestion control.
+
+## Security
+
+### Packet flood incentivization
+
+When TPU/UDP links become congested, packets start to get dropped arbitrarily.
+This incentivizes clients to repeatedly send packets (spam) to increase the chance
+of getting transactions confirmed. Such increase in traffic worsens congestion.
+
+TPU/QUIC implements retransmit and congestion control to address this vulnerability.
+
+### Confidentiality
+
+TPU traffic should be treated as confidential data.
```

Comment:
imo this shouldn't be part of this spec, should focus more on mechanics as you've done above and less on actor behavior

Reply:
Objectively, ignoring if it is appropriate to mention, I believe TPU traffic should be viewed as confidential in nature and is therefore part of a security consideration.

That said, I don't know if it's appropriate to bring up security considerations in this specs repo.
RFC and EIP do, Ethereum specs seem to not.

I do agree that the next section is inappropriate for the spec and will remove it.
MEV is obviously not malicious in nature -- I was more concerned about people negotiating null TLS cipher and getting frontrun by their ISP before it reaches the leader/block builder.

@danpaul000 @t-nelson Any thoughts on the above?

For context, section to be removed:

```diff
-### Confidentiality
-
-TPU traffic should be treated as confidential data.
-
-Transactions are inherently nonpublic before their confirmation on the public blockchain network.
-Exposing transactions to third-parties such as RPC providers or TPU relayers may be detrimental to the signer.
-For example, a malicious TPU participant may intercept, delay, and frontrun token trades to extract arbitrage.
-
-The ideal TPU route consists of a single hop from transaction signer to leader.
-Using TPU/QUIC for end-to-end encryption further reduces MitM risks.
-
-Leaders could always frontrun incoming traffic because they control transaction order in blocks (colloquially known as MEV).
-It is therefore advisable to use MEV-protected applications and directly send sensitive transactions to validators that do not employ MEV.
```

Reply:
Fixed.